### PR TITLE
Add tracepoint example

### PR DESCRIPTION
**Pull request by florianl**
Signed-off-by: Lehner Florian <dev@der-flo.net>

Inspired by issue https://github.com/cilium/ebpf/issues/83 this PR adds an example on how to attach an eBPF program to a tracepoint. It puts the code sample from @mythi into a complete example.

### CODE REVIEW

File: example_program_test.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	ins := asm.Instructions{
+		// store the integer 123 at FP[-8]
+		asm.Mov.Imm(asm.R2, int32(123)),
+		asm.StoreMem(asm.RFP, -8, asm.R2, asm.Word),
+
+		// load registers with arguments for call of FnPerfEventOutput
+		asm.LoadMapPtr(asm.R2, events.FD()),
+		asm.Mov.Reg(asm.R4, asm.RFP),
+		asm.Add.Imm(asm.R4, int32(-8)),
+		asm.LoadImm(asm.R3, 0xffffffff, asm.DWord),
```
Comment (tklauser) [inline example_program_test.go:66]: I'd suggest to initialize these in order `R2-R5`:

```suggestion
		asm.LoadImm(asm.R3, 0xffffffff, asm.DWord),
		asm.Mov.Reg(asm.R4, asm.RFP),
		asm.Add.Imm(asm.R4, int32(-8)),
```

File: example_program_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+				}
+				panic(fmt.Errorf("could not read from reader: %v", err))
+			}
+			fmt.Println(record)
+		}
+	}()
+
+	ins := asm.Instructions{
+		// store the integer 123 at FP[-8]
+		asm.Mov.Imm(asm.R2, int32(123)),
```
Comment (tklauser) [inline example_program_test.go:59]: I don't think the explicit `int32` type conversion is necessar here and also below in all cases where a constant is passed:

```suggestion
		asm.Mov.Imm(asm.R2, 123),
```

File: example_program_test.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		// store the integer 123 at FP[-8]
+		asm.Mov.Imm(asm.R2, 123),
+		asm.StoreMem(asm.RFP, -8, asm.R2, asm.Word),
+
+		// load registers with arguments for call of FnPerfEventOutput
+		asm.LoadMapPtr(asm.R2, events.FD()),
+		asm.LoadImm(asm.R3, 0xffffffff, asm.DWord),
+		asm.Mov.Reg(asm.R4, asm.RFP),
+		asm.Add.Imm(asm.R4, int32(-8)),
+		asm.Mov.Imm(asm.R5, int32(4)),
```
Comment (tklauser) [inline example_program_test.go:67]:
```suggestion
		asm.Add.Imm(asm.R4, -8),
		asm.Mov.Imm(asm.R5, 4),
```

File: example_program_test.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+		// call FnPerfEventOutput
+		asm.FnPerfEventOutput.Call(),
+
+		// set exit code to 0
+		asm.Mov.Imm(asm.R0, int32(0)),
```
Comment (tklauser) [inline example_program_test.go:73]:
```suggestion
		asm.Mov.Imm(asm.R0, 0),
```

File: example_program_test.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		License:      "GPL",
+		Instructions: ins,
+	})
+	if err != nil {
+		panic(fmt.Errorf("could not create new ebpf program: %v", err))
+	}
+	defer prog.Close()
+
+	// tracepoint id from /sys/kernel/debug/tracing/events/syscalls/sys_enter_open/id
+	var tid uint64 = 627
```
[the remark this answers is not in the document]
Comment (mythi) [reply example_program_test.go]: `id` depends on your kernel so it's better to read the content to get the desired id. On my kernel this is 625.

**Comment by tklauser** [comment]
@mythi good point about reading tracepoint `id` from `/sys/kernel/debug/tracing/events/syscalls/sys_enter_open/id`. Do you want to send a PR?

**Comment by mythi** [comment]
@tklauser I'll try to do it

**Pull request merged**